pkg/apis/backupdriver/v1/snapshot.go

```diff
@@ -0,0 +1,148 @@
+// UploadFailed - end state, unable to move to durable storage
+// Canceling - when the SanpshotCancel flag is set, if the Snapshot has not already moved into a terminal state, the
+//             status will move to Canceling.  The snapshot ID will be removed from the status status if has been filled in
+//             and the snapshot ID will not longer be valid for a Clone operation
+// Canceled - the operation was canceled, the snapshot ID is not valid
+type SnapshotPhase string
+
+const (
+	SnapshotPhaseNew            SnapshotPhase = "New"
+	SnapshotPhaseInProgress     SnapshotPhase = "InProgress"
+	SnapshotPhaseSnapshotted    SnapshotPhase = "Snapshotted"
+	SnapshotPhaseSnapshotFailed SnapshotPhase = "SnapshotFailed"
+	SnapshotPhaseUploading      SnapshotPhase = "Uploading"
+	SnapshotPhaseUploaded       SnapshotPhase = "Uploaded"
+	SnapshotPhaseUploadFailed   SnapshotPhase = "UploadFailed"
+	SnapshotPhaseCanceling      SnapshotPhase = "Canceling"
+	SnapshotPhaseCanceled       SnapshotPhase = "Canceled"
+)
+
+// UploadOperationProgress represents the progress of a
+// Upload operation
+type SnapshotProgress struct {
+	// +optional
+	TotalBytes int64 `json:"totalBytes,omitempty"`
+
+	// +optional
+	BytesDone int64 `json:"bytesDone,omitempty"`
+}
+
+type SnapshotStatus struct {
+	// Phase is the current state of the Upload.
+	Phase SnapshotPhase `json:"phase,omitempty"`
+
+	// Message is a message about the snapshot's status.
+	// +optional
+	Message string `json:"message,omitempty"`
+
+	// Progress for the upload
+	Progress SnapshotProgress `json:"progress, omitempty"`
+
+	// Snapshot ID that has been taken.  This will be filled in when the phase goes to "Snapshotted"
+	// +optional
+	SnapshotID string `json:"snapshotID, omitempty"`
+
+	// Metadata for the snapshotted object
+	Metadata []byte `"json:metadata, omitempty""`
+}
+
+/*
+ Snapshot is used to request that a snapshot is taken.  It is not used to manage the inventory of snapshots and does not
+ need to exist in order to clone the resource from a snapshot
+*/
+type Snapshot struct {
+	// TypeMeta is the metadata for the resource, like kind and apiversion
+	meta_v1.TypeMeta `json:",inline"`
+
+	// +optional
+	meta_v1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the custom resource spec
+	Spec SnapshotSpec `json:"spec"`
+
+	// Current status of the snapshot operation
+	// +optional
+	Status SnapshotStatus `json:"status,omitempty"`
+
+	// SnapshotCancel indicates request to cancel ongoing snapshot.  SnapshotCancel can be set at anytime before
+	// the snapshot reaches a terminal phase.  If the snapshot has reached a terminal phase
+	SnapshotCancel bool `json:"snapshotCancel,omitempty"`
+}
+
+type CloneSpec struct {
+	SnapshotID string `json:"snapshotID"`
+
+	// +optional - if set, this overrides metadata that was stored in the snapshot
+	Metadata     []byte `json:"metadata, omitempty"`
+	DesiredClass string `json:"desiredClass"`
```

Comment:
Is DesiredClass required?
Does this refer to a new BackupClass?  I don't see it is defined here.  Please add comments to clarify what DesiredClass means and what type it is.